### PR TITLE
Read mobtypes.txt and handle standard and custom creatures animations

### PR DESCRIPTION
**Pull request by DavideRei**
Read mobtypes.txt (if available among mul files) and cache it to check animations type when using creatures animations.
Now standard anims (SEA_MONSTER, ANIMAL, MONSTER) and UOP anims are handled correctly depending on which actions anim they have on anim files.
About the anims, i checked client latest version and writed down the info here: https://docs.google.com/spreadsheets/d/1kXHctxuHJIuv6yvsA_nzjov_A1IAYkeKpTqx5_c8yM8/edit?usp=sharing

Added alert anim on NPCs when they start combat, pillage anim when they loot something (items on ground or in bodies) and "summon" anim when they are summoned or added. 
I'm still dubious about the SetTimeout command after summon action. 
I added it to show the entire animation, because if the creature moves, the animation is interrupted. 
Some monsters have 2-3 sec. animation but the timeout freezes them until is completed, so for example if a player summon a creature he can't command it until the timeout is completed. 
What are your thoughts about this? Is there a bettere way to do it?

Now custom anims can be handled by scripts using ANIM command (It was almost useless before, so i decided to replace it) under CHARDEF, to set which anim actions they have/don't have, using the following flags.

MONSTER (H slot) default configuration is with actions: 0 (Walk), 1 (Stand), 2 (Die1), 3 (Die2), 4 (Attack1), 5 (Attack2), 6 (Attack3), 10 (GetHit), 11 (Pillage), 15 (Block Right), 16 (Block Left), 17 (Idle), 18 (Fidget)
    No DIE2 action		= 01,
    No ATTACK1 action   = 02,
    No ATTACK2 action   = 04,
    No ATTACK3 action   = 08,
    No GETHIT action    = 010,
    No PILLAGE action   = 020,
    STOMP action		= 040,
    CAST2 action        = 080,
    CAST3 action        = 0100,
    No BLOCKR action    = 0200,
    No BLOCKL action    = 0400,
    No FIDGET1 action   = 0800,
    No FIDGET2 action   = 01000,
    FLY action          = 02000,
    Anims with Attacks 1-2-3 in actions 9,10,11 and with GetHit in action 20 (there are many custom WOW anims with this configuration) = 0x10000

ANIMAL (L slot) default configuration is with all actions: 1 - 12
    No EAT action       = 01,
    No ALERT action     = 02,
    No ATTACK2 action   = 04,
    No GETHIT action    = 08,
    No FIDGET1 action   = 010,
    No FIDGET2 action   = 020,
    No LIEDOWN action   = 040,
    No DIE2 action      = 080

So for example if you have a MONSTER (H slot) flying custom anim, with Cast2 action but not Pillage action you can set ANIM=020A0
If you have a custom anim with default configuration, you don't need to set ANIM= (i checked with my custom anims to use the most suitable config to don't have to change all chardefs)

Issue: #885

### CODE REVIEW

**Comment by Jhobean** [comment]
Good job!  Just want to link the related issue: https://github.com/Sphereserver/Source-X/issues/885

File: src/game/uo_files/CUOMobtypes.cpp [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+
+    CUOMobTypesType mobTypesRow = {0,0};
+
+    _mobTypesEntries.clear();
+
+    CSFileText csvMobTypes;
+
+    if (g_Install.OpenFile(csvMobTypes, "mobtypes.txt", (word)(OF_READ | OF_TEXT | OF_DEFAULTMODE)))
+    {
+        _mobTypesEntries.resize(4096);
```
Comment (cbnolok) [inline src/game/uo_files/CUOMobtypes.cpp:26]: Is 4096 an arbitrary number or some hardcoded client limit?

File: src/game/uo_files/CUOMobtypes.cpp [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+
+                int len = (int)tmpString.length();
+                len = Str_TrimEndWhitespace(tmpString.data(), len);
+
+                if (len == 0 || tmpString[0] == '#') //Empty line or commented
+                    continue;
+
+                //Split the string
+                char* pch;
+                pch = strtok(tmpString.data(), " \t");
```
Comment (cbnolok) [inline src/game/uo_files/CUOMobtypes.cpp:55]: You could check, if you like so, Str_Parse. It's a Sphere function which hasn't the drawbacks of strtok, mainly it's not thread safe.

File: src/game/uo_files/CUOMobtypes.h [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+#define PACK_NEEDED
+#else
+    // GCC based compiler you can add:
+#define PACK_NEEDED __attribute__ ((packed))
+#endif
+
+/**
+* mobtypes.txt
+*/
+struct CUOMobTypesType
```
Comment (cbnolok) [inline src/game/uo_files/CUOMobtypes.h:25]: You could simplify this as we don't need structure packing here. It's usually needed when we want to read a sequence of bytes (from a UO mul file) and with those fill an entire struct. Structure packing instructs the compiler to not add padding bytes.
Pro read for more detail: http://www.catb.org/esr/structure-packing/
Faster answer: https://stackoverflow.com/questions/4306186/structure-padding-and-packing

**Review by cbnolok — COMMENTED** [review]
That's an excellent one, thanks!
I'd suggest some little changes before merging it. One of them would be to add a ini setting to enable/disable mobtypes loading, so giving the choice to use the old behavior

File: src/game/uo_files/CUOMobtypes.cpp (hunk 2 above)
Comment (DavideRei) [reply src/game/uo_files/CUOMobtypes.cpp:55]: Like this?
```
//Split the string
std::vector<tchar*> splitArray;
splitArray.resize(3);
size_t iQty = Str_ParseCmds(tmpString.data(), splitArray.data(), 3, " \t#");
```

**Comment by DavideRei** [comment]
> That's an excellent one, thanks! I'd suggest some little changes before merging it. One of them would be to add a ini setting to enable/disable mobtypes loading, so giving the choice to use the old behavior

About the ini setting, I thought it wasn't required because the mobtypes file is not loaded (and in this case Sphere uses the old behavior to distinguish between monster and animal) unless it's in the mul folder.
Therefore if someone doesn't want to load it, he could simply remove it. Usually if you are running Sphere on a server you don't need the full installation but only few files: hues, map, multi, radarcol, statics and tiledata.
In addition, if you have the mobtypes file why not use it?

File: src/game/uo_files/CUOMobtypes.cpp (hunk 2 above)
Comment (cbnolok) [reply src/game/uo_files/CUOMobtypes.cpp:55]: Yes, or you could just use tchar* ppColumns[3] in place of that vector

File: src/game/uo_files/CUOMobtypes.h (hunk 3 above)
Comment (cbnolok) [reply src/game/uo_files/CUOMobtypes.h:25]: Exactly

**Comment by cbnolok** [comment]
One could simply just want to use their full client folder also for sphere. In any case, i feel that an ini setting is more straightforward and simple to manage for people wishing to keep the old "buggy" behavior just because it's "classic sphere".

**Comment by DavideRei** [comment]
Done but how can I resolve merge conflicts now?

**Comment by cbnolok** [comment]
Manually merged, needed in order to fix conflicts.

**Pull request closed without merging**